lib/segment/src/vector_storage/chunked_mmap_vectors.rs

```diff
@@ -0,0 +1,195 @@
+}
+
+#[derive(Serialize, Deserialize)]
+struct ChunkedMMapConfig {
+    chunk_size: usize,
+    dim: usize,
+}
+
+pub struct ChunkedMMapVectors<T> {
+    config: ChunkedMMapConfig,
+    status: ChunkedMMapStatus,
+    _phantom: PhantomData<T>,
+    chunks: Vec<MmapMut>,
+    status_mmap: MmapMut,
+    directory: PathBuf,
+}
+
+impl ChunkedMMapVectors<VectorElementType> {
+    fn status_file(directory: &Path) -> PathBuf {
+        directory.join(STATUS_FILE_NAME)
+    }
+
+    fn write_to_mmap<T>(mmap: &mut MmapMut, offset: usize, data: T) {
+        debug_assert!(offset + size_of::<T>() <= mmap.len());
+        let ptr = mmap.as_mut_ptr();
+        unsafe {
+            let byte_data = transmute_to_u8(&data);
+            std::ptr::copy_nonoverlapping(byte_data.as_ptr(), ptr.add(offset), byte_data.len());
+        }
+    }
+
+    fn write_array_to_mmap<T>(mmap: &mut MmapMut, offset: usize, data: &[T]) {
+        debug_assert!(offset + data.len() * size_of::<T>() <= mmap.len());
+        let ptr = mmap.as_mut_ptr();
+        unsafe {
+            let byte_data = transmute_to_u8_array(data);
+            std::ptr::copy_nonoverlapping(byte_data.as_ptr(), ptr.add(offset), byte_data.len());
+        }
+    }
+
+    fn read_from_mmap<T>(mmap: &MmapMut, offset: usize) -> &T {
+        debug_assert!(offset + size_of::<T>() <= mmap.len());
+        let ptr = mmap.as_ptr();
+        unsafe { &*(ptr.add(offset) as *const T) }
+    }
+
+    fn read_array_from_mmap<T>(mmap: &MmapMut, offset: usize, len: usize) -> &[T] {
+        debug_assert!(offset + len * size_of::<T>() <= mmap.len());
+        let byte_len = len * size_of::<T>();
+        let slice = &mmap[offset..offset + byte_len];
+        transmute_from_u8_to_array(slice)
+    }
+
+    fn ensure_status_file(directory: &Path) -> OperationResult<MmapMut> {
+        let status_file = Self::status_file(directory);
+        if !status_file.exists() {
+            {
+                let length = std::mem::size_of::<usize>() as u64;
+                let file = OpenOptions::new()
+                    .create(true)
+                    .write(true)
+                    .open(&status_file)?;
+                file.set_len(length)?;
+            }
+            let mut mmap = open_write_mmap(&status_file)?;
+            Self::write_to_mmap(&mut mmap, 0, 0usize);
+            mmap.flush()?;
+            Ok(mmap)
+        } else {
+            open_write_mmap(&status_file)
+        }
+    }
+
+    fn ensure_config(directory: &Path, dim: usize) -> OperationResult<ChunkedMMapConfig> {
+        let config_file = directory.join(CONFIG_FILE_NAME);
+        if !config_file.exists() {
+            let config = ChunkedMMapConfig {
+                chunk_size: DEFAULT_CHUNK_SIZE,
+                dim,
+            };
+            let mut file = OpenOptions::new()
+                .create(true)
+                .write(true)
+                .open(&config_file)?;
+            serde_json::to_writer_pretty(&mut file, &config)?;
```

Comment:
Prettified JSON is larger and takes extra time. Not really needed I think. Human readers can prettify in their editor.